### ISSUE TITLE
When using flow loss, the training time increases significantly in the 4D reconstruction task

### ISSUE THREAD

**Issue by tb2-sy** [Zerg-Overmind/GaussianFlow#8]
Thanks for your nice work! 
I tried to integrate flow loss into [4D-GS](https://github.com/hustvl/4DGaussians), but found that the training time increased significantly when executing the loss.backward() operation, from the original one hour to the current estimated training time of 100 hours, which is almost 100 times longer. Is this reasonable?

**Comment by Zerg-Overmind** [comment]
Hi, training time should increase but definitely not that slow in our experiments. Could you please let me know the resolution, the number of K per image and your batch size?

**Comment by tb2-sy** [comment]
The resolution is 1920*1080, the batch size is 1, and K is the default 20.

**Comment by Zerg-Overmind** [comment]
superized...but thanks for letting me know and I will try the code anyway..

**Comment by tb2-sy** [comment]
Is this because the resolution is too high?

**Comment by Zerg-Overmind** [comment]
I don't think so since we did experiments on dynerf dataset whose resolution is also 1k or higher.

**Comment by tb2-sy** [comment]
Okay, this is a bit weird.

**Comment by tb2-sy** [comment]
![43731717429958_ pic](https://github.com/Zerg-Overmind/GaussianFlow/assets/61854446/8dbd0725-bbcf-4f43-af18-3278002cc4c5)
Checking the time consumed by each part of the backward process, we found some problems. The backward process with flow loss consumes 30+ seconds. And between [https://github.com/Zerg-Overmind/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L419](url) to line 448 (Time for RasterizeGaussiansBackwardCUDA2). The time consumption is thousands of times that of https://github.com/Zerg-Overmind/diff-gaussian-rasterization/blob/main/cuda_rasterizer/backward.cu#L756 to line 783 (Time for renderCUDA). As shown in the figure above, when flow loss is involved, cuda backward will be executed twice. The difference between the two times when flow loss is involved is very large. Why is this?

**Comment by Xharlie** [comment]
30 times is crazy btw, i think there must be something wrong, forgive us not being able to debug on your server, we can only verify everything on our end. Maybe the cuda/gpu version is very different?  30 times sounds like gpu is not properly used.

**Comment by penghaohsu** [comment]
Hi @tb2-sy , I'm trying to do the same thing to integrate the flow loss to 4DGS, but the flow I rendered looks quite weird. Did you successfully integrate it into 4DGS codebase?

**Comment by zdrac** [comment]
> Hi @tb2-sy , @penghaohsu0222 ,  I'm also trying to integrate the flow loss to 4DGS. Have you solved the problem of slow training on 4DGS?

**Comment by 120829** [comment]
> > Hi @tb2-sy , @penghaohsu0222 ,  I'm also trying to integrate the flow loss to 4DGS. Have you solved the problem of slow training on 4DGS?你好，，我也在尝试将流量损失集成到 4DGS 中。 4DGS训练慢的问题解决了吗？

Hello, have you solved this problem? I'm also having the same issue.

**Comment by Mercerai** [comment]
have you ever solved this issue? I also encounter the same problem. When adopting the flow loss. the backward has become very very slow.   @Zerg-Overmind

**Comment by LabubuJ** [comment]
> > > Hi @tb2-sy , @penghaohsu0222 ,  I'm also trying to integrate the flow loss to 4DGS. Have you solved the problem of slow training on 4DGS?你好，，我也在尝试将流量损失集成到 4DGS 中。 4DGS训练慢的问题解决了吗？
> 
> Hello, have you solved this problem? I'm also having the same issue.

@120829 Have you solved this problem? I have a try and find the same question.

**Comment by Langwenchong** [comment]
Facing the same issue — when the flow renderer is included in the training backward pass, the training time increases significantly. Based on the Profiler output, it seems that the main bottleneck is related to the IndexBackward0 operator. Do you have any insights or progressive understanding about this?

<img width="1399" height="439" alt="Image" src="https://github.com/user-attachments/assets/552282c2-7300-4cec-a9ba-5270afae86cb" />
